include/vara/Feature/ConstraintParser.h

```diff
@@ -0,0 +1,353 @@
+#ifndef VARA_FEATURE_CONSTRAINTPARSER_H
+#define VARA_FEATURE_CONSTRAINTPARSER_H
+
+#include "vara/Feature/Constraint.h"
+#include "vara/Feature/Feature.h"
+
+#include <iostream>
+
+namespace vara::feature {
+
+//===----------------------------------------------------------------------===//
+//                               ConstraintToken
+//===----------------------------------------------------------------------===//
+
+class ConstraintToken {
+public:
+  using Precedence = int;
+
+  enum class ConstraintTokenKind {
+    ERROR,
+    END_OF_FILE,
+    WHITESPACE,
+    IDENTIFIER,
+    NUMBER,
+    L_PAR,
+    R_PAR,
+    NOT,
+    OR,
+    AND,
+    EQUAL,
+    NOT_EQUAL,
+    LESS,
+    GREATER,
+    IMPLIES,
+    EQUIVALENT,
+    PLUS,
+    MINUS,
+    STAR,
+    NEG
+  };
+
+  ConstraintToken(ConstraintTokenKind Kind) : Kind(Kind) {}
+  ConstraintToken(ConstraintTokenKind Kind, std::string Value)
+      : Kind(Kind), Value(Value) {}
+  ConstraintToken(const ConstraintToken &) = delete;
+  ConstraintToken &operator=(ConstraintToken &) = delete;
+  ConstraintToken(ConstraintToken &&) = delete;
+  ConstraintToken &operator=(ConstraintToken &&) = delete;
+  virtual ~ConstraintToken() = default;
+
+  [[nodiscard]] ConstraintTokenKind getKind() const { return Kind; };
+
+  [[nodiscard]] std::optional<std::string> getValue() { return Value; }
+
+  [[nodiscard]] Precedence calcPrecedence() const {
+    switch (Kind) {
+    case ConstraintTokenKind::STAR:
+      return 1;
+    case ConstraintTokenKind::PLUS:
+    case ConstraintTokenKind::MINUS:
+      return 2;
+    case ConstraintTokenKind::LESS:
+    case ConstraintTokenKind::GREATER:
+      return 3;
+    case ConstraintTokenKind::EQUAL:
+    case ConstraintTokenKind::NOT_EQUAL:
+      return 4;
+    case ConstraintTokenKind::AND:
+      return 5;
+    case ConstraintTokenKind::OR:
+      return 6;
+    default:
+      return 7;
+    }
+  }
+
+private:
+  const ConstraintTokenKind Kind;
+  const std::optional<std::string> Value{std::nullopt};
+};
+
+//===----------------------------------------------------------------------===//
+//                               ConstraintToken
+//===----------------------------------------------------------------------===//
+
+class ConstraintLexer {
+public:
+  using TokenListTy = std::vector<std::unique_ptr<ConstraintToken>>;
```

Comment:
```suggestion
  using TokenListTy = std::vector<ConstraintToken>;
```
Wo we really need a pointer indirection here? CTokens seem quite small, so a few copies/moves should not hurt.